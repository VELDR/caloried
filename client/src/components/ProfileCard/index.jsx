import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
import { FormattedMessage, injectIntl } from 'react-intl';
import { Avatar } from '@mui/material';
import { Balance, KeyboardDoubleArrowDown, KeyboardDoubleArrowUp } from '@mui/icons-material';

import classes from './style.module.scss';

const ProfileCard = ({ user, intl: { formatMessage } }) => {
  const navigate = useNavigate();
  const getGoalInfo = (goal) => {
    switch (goal) {
      case 'gain':
        return {
          icon: <KeyboardDoubleArrowUp />,
          text: formatMessage({ id: 'app_gaining_weight' }),
          className: classes.gainColor,
        };
      case 'maintain':
        return {
          icon: <Balance />,
          text: formatMessage({ id: 'app_maintaining_weight' }),
          className: classes.maintainColor,
        };
      case 'lose':
        return {
          icon: <KeyboardDoubleArrowDown />,
          text: formatMessage({ id: 'app_losing_weight' }),
          className: classes.loseColor,
        };
      default:
        return { icon: null, text: '', className: '' };
    }
  };

  const goalContent = getGoalInfo(user?.goal);

  return (
    <div className={classes.profile}>
      <div className={classes.profile__header}>
        <Avatar src={`${import.meta.env.VITE_API_BASE_URL}${user?.avatar}`} className={classes.avatar} />
        <div className={classes.identity}>
          <div className={classes.username}>{user?.username}</div>
          <div className={classes.email}>{user?.email}</div>
          <div className={classes.button} onClick={() => navigate('/profile')}>
            <FormattedMessage id="app_view_profile" />
          </div>
        </div>
      </div>
      <div className={classes.profile__footer}>
        <div className={classes.intake}>
          <div className={classes.intake__label}>
            <FormattedMessage id="app_calories" />
          </div>
          <div className={classes.intake__value}>
            {user?.bmr}{' '}
            <span>
              kcal/
              <FormattedMessage id="app_day" />
            </span>
          </div>
        </div>
        <div className={classes.intake}>
          <div className={classes.intake__label}>
            <FormattedMessage id="app_protein" />
          </div>
          <div className={classes.intake__value}>
            {user?.proteinIntake}{' '}
            <span>
              g/
              <FormattedMessage id="app_day" />
            </span>
          </div>
        </div>
        <div className={classes.intake}>
          <div className={classes.intake__label}>
            <FormattedMessage id="app_carbs" />
          </div>
          <div className={classes.intake__value}>
            {user?.carbsIntake}{' '}
            <span>
              g/
              <FormattedMessage id="app_day" />
            </span>
          </div>
        </div>
        <div className={classes.intake}>
          <div className={classes.intake__label}>
            <FormattedMessage id="app_fat" />
          </div>
          <div className={classes.intake__value}>
            {user?.fatIntake}{' '}
            <span>
              g/
              <FormattedMessage id="app_day" />
            </span>
          </div>
        </div>
      </div>
      <div className={`${classes.goal} ${goalContent.className}`}>
        {goalContent.icon} {goalContent.text}
      </div>
    </div>
  );
};

ProfileCard.propTypes = {
  user: PropTypes.object,
  intl: PropTypes.object,
};

export default injectIntl(ProfileCard);
