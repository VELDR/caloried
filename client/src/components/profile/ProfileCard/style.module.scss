@import '@styles/base';

.profile {
  background-color: var(--color-bg-secondary);
  padding: 1rem;
  border-radius: 1rem;
  position: relative;
  overflow: hidden;
  @include flex-column(1rem);
  @include apply-at-md-size {
    width: 50%;
  }
  @include apply-at-lg-size {
    width: 60%;
  }
  &__header {
    @include flex-items-center(1rem);
    .avatar {
      @include box(6rem);
    }
    .identity {
      @include flex-column;
      .username {
        font-size: $font-size-large;
        font-weight: 500;
      }
      .email {
        color: var(--color-text-secondary);
        @include truncate(15rem);
        font-size: $font-size-regular;
        @include apply-at-xs2-size {
          max-width: 100%;
        }
        @include apply-at-md-size {
          max-width: 12rem;
        }
        @include apply-at-lg-size {
          max-width: 100%;
        }
      }
      .button {
        @include primary-button;
        width: fit-content;
        font-size: $font-size-small;
        margin-top: 0.5rem;
      }
    }
  }
  &__footer {
    @include flex-row(1rem);
    justify-content: space-between;
    .intake {
      &__label {
        font-size: $font-size-smaller;
        text-transform: uppercase;
        color: var(--color-text-secondary);
        font-weight: 500;
      }
      &__value {
        font-weight: 500;
        span {
          font-weight: 300;
          font-size: $font-size-smaller;
        }
      }
    }
  }
  .goal {
    position: absolute;
    top: 0;
    right: 0;
    padding: 0.25rem 0.75rem;
    text-transform: capitalize;
    @include flex-items-center(0.25rem);
    font-size: $font-size-small;
    border-bottom-left-radius: 1.5rem;
    background-color: var(--color-bg-primary);

    svg {
      @include box(1rem);
      animation: shake 1.5s ease-in-out infinite;
    }
  }
  .gainColor {
    color: var(--color-primary);
  }
  .maintainColor {
    color: $color-fat;
  }
  .loseColor {
    color: $color-error;
  }
}

@keyframes shake {
  0%,
  100% {
    transform: rotate(0deg);
  }

  5% {
    transform: rotate(6deg);
  }

  10%,
  20% {
    transform: rotate(-6deg);
  }

  15%,
  25%,
  50% {
    transform: rotate(6deg);
  }

  30%,
  100% {
    transform: rotate(0deg);
  }
}
