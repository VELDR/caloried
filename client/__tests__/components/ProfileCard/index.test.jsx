import ProfileCard from '@components/profile/ProfileCard';
import { fireEvent, render } from '@utils/testHelper';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

const mockProps = {
  intl: {
    formatMessage: jest.fn(),
  },
};

const renderProfileCard = (user) => {
  const wrapper = render(<ProfileCard user={user} {...mockProps} />);
  return { wrapper, ...wrapper };
};

describe('ProfileCard Component', () => {
  const testUser = {
    avatar: 'avatar-url',
    username: 'testuser',
    email: 'test@example.com',
    bmr: 2000,
    proteinIntake: 100,
    carbsIntake: 150,
    fatIntake: 50,
  };

  const goals = [
    { goal: 'gain', expectedText: 'Menaikkan Berat Badan' },
    { goal: 'maintain', expectedText: 'Mempertahankan Berat Badan' },
    { goal: 'lose', expectedText: 'Menurunkan Berat Badan' },
  ];

  goals.forEach(({ goal, expectedText }) => {
    it(`should render component correctly for "${goal}" goal`, () => {
      const user = { ...testUser, goal };
      const { getByText, getByTestId } = renderProfileCard(user);

      expect(getByTestId('profile-card')).toBeInTheDocument();
      expect(getByText('testuser')).toBeInTheDocument();
      expect(getByText('test@example.com')).toBeInTheDocument();
      expect(getByText(expectedText)).toBeInTheDocument();
    });
  });

  it('should navigate to profile page on button click', () => {
    const user = { ...testUser, goal: 'lose' };
    const { getByTestId } = renderProfileCard(user);

    const viewProfileButton = getByTestId('view-profile-button');
    fireEvent.click(viewProfileButton);

    expect(mockNavigate).toHaveBeenCalledWith(`/profile`);
  });

  it('should match with snapshot', () => {
    const user = { ...testUser, goal: 'lose' };
    const { wrapper } = renderProfileCard(user);

    expect(wrapper).toMatchSnapshot();
  });
});
