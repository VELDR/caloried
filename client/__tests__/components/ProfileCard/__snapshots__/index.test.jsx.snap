// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileCard Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="profile"
        data-testid="profile-card"
      >
        <div
          class="profile__header"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular avatar css-1wlk0hk-MuiAvatar-root"
          >
            <img
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="http://localhost:8080avatar-url"
            />
          </div>
          <div
            class="identity"
          >
            <div
              class="username"
            >
              testuser
            </div>
            <div
              class="email"
            >
              test@example.com
            </div>
            <div
              class="button"
              data-testid="view-profile-button"
            >
              Lihat Profil
            </div>
          </div>
        </div>
        <div
          class="profile__footer"
        >
          <div
            class="intake"
          >
            <div
              class="intake__label"
            >
              Kalori
            </div>
            <div
              class="intake__value"
            >
              2000
               
              <span>
                kcal/
                hari
              </span>
            </div>
          </div>
          <div
            class="intake"
          >
            <div
              class="intake__label"
            >
              Protein
            </div>
            <div
              class="intake__value"
            >
              100
               
              <span>
                g/
                hari
              </span>
            </div>
          </div>
          <div
            class="intake"
          >
            <div
              class="intake__label"
            >
              Karbohidrat
            </div>
            <div
              class="intake__value"
            >
              150
               
              <span>
                g/
                hari
              </span>
            </div>
          </div>
          <div
            class="intake"
          >
            <div
              class="intake__label"
            >
              Lemak
            </div>
            <div
              class="intake__value"
            >
              50
               
              <span>
                g/
                hari
              </span>
            </div>
          </div>
        </div>
        <div
          class="goal loseColor"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardDoubleArrowDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 6.41 16.59 5 12 9.58 7.41 5 6 6.41l6 6z"
            />
            <path
              d="m18 13-1.41-1.41L12 16.17l-4.59-4.58L6 13l6 6z"
            />
          </svg>
           
          Menurunkan Berat Badan
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="profile"
      data-testid="profile-card"
    >
      <div
        class="profile__header"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular avatar css-1wlk0hk-MuiAvatar-root"
        >
          <img
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="http://localhost:8080avatar-url"
          />
        </div>
        <div
          class="identity"
        >
          <div
            class="username"
          >
            testuser
          </div>
          <div
            class="email"
          >
            test@example.com
          </div>
          <div
            class="button"
            data-testid="view-profile-button"
          >
            Lihat Profil
          </div>
        </div>
      </div>
      <div
        class="profile__footer"
      >
        <div
          class="intake"
        >
          <div
            class="intake__label"
          >
            Kalori
          </div>
          <div
            class="intake__value"
          >
            2000
             
            <span>
              kcal/
              hari
            </span>
          </div>
        </div>
        <div
          class="intake"
        >
          <div
            class="intake__label"
          >
            Protein
          </div>
          <div
            class="intake__value"
          >
            100
             
            <span>
              g/
              hari
            </span>
          </div>
        </div>
        <div
          class="intake"
        >
          <div
            class="intake__label"
          >
            Karbohidrat
          </div>
          <div
            class="intake__value"
          >
            150
             
            <span>
              g/
              hari
            </span>
          </div>
        </div>
        <div
          class="intake"
        >
          <div
            class="intake__label"
          >
            Lemak
          </div>
          <div
            class="intake__value"
          >
            50
             
            <span>
              g/
              hari
            </span>
          </div>
        </div>
      </div>
      <div
        class="goal loseColor"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="KeyboardDoubleArrowDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 6.41 16.59 5 12 9.58 7.41 5 6 6.41l6 6z"
          />
          <path
            d="m18 13-1.41-1.41L12 16.17l-4.59-4.58L6 13l6 6z"
          />
        </svg>
         
        Menurunkan Berat Badan
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
